### PR TITLE
Axis Aligned Bounding Box

### PR DESCRIPTION
**Pull request by steviedale**
This pull request addresses the need for a function that segments looped paths along vectors diagonal to the major and middle axis of the path.

### CODE REVIEW

**Comment by colin-lewis-19** [comment]
We should change terminology from 'segmenting' to 'splitting' to prevent conflating the splitting of edge paths with segmentation of surfaces

**Comment by colin-lewis-19** [comment]
I opened a PR against your branch @steviedale that should ensure it works generally. When that is merged, I believe this will be good to go @DavidMerzJr

**Comment by steviedale** [comment]
@colin-lewis-19 @DavidMerzJr:
I just merged @colin-lewis-19's PR and renamed `segmentByAxes` to `splitByAxes`. Should be good to go now.

File: tool_path_planner/CMakeLists.txt [hunk 1]
```diff
@@ -81,10 +83,19 @@ add_dependencies(${PROJECT_NAME}
   ${catkin_EXPORTED_TARGETS}
 )
 
+add_executable(test_segment_axes
+  src/test_segment_axes.cpp
```
Comment (DavidMerzJr) [inline tool_path_planner/CMakeLists.txt]: Test files should go in the /test/ directory, and should be built as tests, not executables.

File: tool_path_planner/src/utilities.cpp [hunk 2]
```diff
@@ -10,6 +10,7 @@
 #include <limits>
 #include <cmath>
 #include <numeric>
+#include <algorithm> // std::sort
```
Comment (DavidMerzJr) [inline tool_path_planner/src/utilities.cpp:13]: Thank you for marking what needs the std header - it's actually good to know.

File: tool_path_planner/src/test_segment_axes.cpp [hunk 3]
```diff
@@ -0,0 +1,271 @@
+#include <ros/ros.h>
```
Comment (DavidMerzJr) [inline tool_path_planner/src/test_segment_axes.cpp]: I'm not sure this file should be added to the noether repository - it is not a true unit test, and is not integrated with / does not match the style of the existing tests.  It looks more like a piece of code you used to bootstrap test whether the new functions were working.  If you integrate to utest.cpp, I'd be alright merging it.

File: tool_path_planner/src/utilities.cpp [hunk 4]
```diff
@@ -484,6 +503,165 @@ ToolPaths addExtraPaths(const ToolPaths& tool_paths, double offset_distance)
...
 }
 
+ToolPath splitByAxes(const ToolPathSegment& tool_path_segment)
+{
+  using Point = pcl::PointXYZ;
+  // Sanity check - tool path segment must not be empty
+  if (tool_path_segment.size() == 0)
+  {
+    ROS_ERROR_STREAM("Tool path segment 0 is empty.");
+    // TODO: throw an exception
```
Comment (DavidMerzJr) [inline tool_path_planner/src/utilities.cpp]: This does not seem worthy of an exception (which you didn't add anyway).  If a segment is of zero length, just immediately return an empty ToolPath.

File: tool_path_planner/src/utilities.cpp [hunk 5]
```diff
@@ -484,6 +503,165 @@ ToolPaths addExtraPaths(const ToolPaths& tool_paths, double offset_distance)
...
+  return splitByAxes(tool_path_segment, major_axis, perp_axis);
+}
+
+ToolPath splitByAxes(const ToolPathSegment& tool_path_segment, const Eigen::Vector3f& axis_1, const Eigen::Vector3f& axis_2)
+{
+  // Sanity check - tool path segment must not be empty
+  if (tool_path_segment.size() == 0)
+  {
+    ROS_ERROR_STREAM("Tool path segment 0 is empty.");
+    // TODO: throw an exception
```
Comment (DavidMerzJr) [inline tool_path_planner/src/utilities.cpp]: Again, this is not worthy of an exception - just return an empty ToolPath and let the calling code decide what to do with it.

File: tool_path_planner/src/utilities.cpp [hunk 6]
```diff
@@ -484,6 +503,165 @@ ToolPaths addExtraPaths(const ToolPaths& tool_paths, double offset_distance)
...
+  }
+  return new_tool_path;
+}
+
+ToolPaths splitByAxes(const ToolPaths& tool_paths)
+{
+  if (tool_paths.size() == 0)
+  {
+    ROS_ERROR_STREAM("Tool paths is empty.");
+    // TODO: throw an exception
```
Comment (DavidMerzJr) [inline tool_path_planner/src/utilities.cpp]: This is still not really an exception.  What is zero divided in half? Still zero.  If you get something empty, just return something empty.

File: tool_path_planner/src/utilities.cpp [hunk 7]
```diff
@@ -484,6 +503,165 @@ ToolPaths addExtraPaths(const ToolPaths& tool_paths, double offset_distance)
...
+  }
+
+  using Point = pcl::PointXYZ;
```
Comment (DavidMerzJr) [inline tool_path_planner/src/utilities.cpp]: Avoid using declarations, as they make the code more difficult to inspect - now someone has to go find this using declaration to understand what type you are using.  It saves you time once when initially writing but costs someone time every time they have to read the code.

File: tool_path_planner/src/utilities.cpp [hunk 8]
```diff
@@ -484,6 +503,165 @@ ToolPaths addExtraPaths(const ToolPaths& tool_paths, double offset_distance)
...
+
+  for (std::size_t tool_path_index = 0; tool_path_index < tool_paths.size(); ++tool_path_index)
+  {
+    ToolPath tool_path = tool_paths[tool_path_index];
+    // Sanity check - tool path must have exactly one segment
+    if (tool_path.size() != 1)
+    {
+      ROS_ERROR_STREAM("Tool path " << tool_path_index << " contains " << tool_path.size() <<
+                       " segments. Expected exactly 1.");
+      // TODO: throw an exception
```
Comment (DavidMerzJr) [inline tool_path_planner/src/utilities.cpp]: Why must the tool_path have exactly one segment?  Couldn't we just iteratively call split() on each segment it contains?

File: tool_path_planner/src/utilities.cpp [hunk 9]
```diff
@@ -484,6 +503,165 @@ ToolPaths addExtraPaths(const ToolPaths& tool_paths, double offset_distance)
...
+  }
+  return new_tool_paths;
+}
+
+ToolPaths splitByAxes(const ToolPaths& tool_paths, const Eigen::Vector3f& axis_1, const Eigen::Vector3f& axis_2)
+{
+  if (tool_paths.size() == 0)
+  {
+    ROS_ERROR_STREAM("Tool paths is empty.");
+    // TODO: throw an exception
```
Comment (DavidMerzJr) [inline tool_path_planner/src/utilities.cpp]: Still don't like these "TODO: Exception".  Don't like the unresolved TODOs, don't like that the proposed solution is an exception.  Just return something empty.

File: tool_path_planner/src/utilities.cpp [hunk 10]
```diff
@@ -484,6 +503,165 @@ ToolPaths addExtraPaths(const ToolPaths& tool_paths, double offset_distance)
...
+  }
+
+  using Point = pcl::PointXYZ;
```
Comment (DavidMerzJr) [inline tool_path_planner/src/utilities.cpp]: Please remove the using statements.

File: tool_path_planner/src/utilities.cpp [hunk 11]
```diff
@@ -484,6 +503,165 @@ ToolPaths addExtraPaths(const ToolPaths& tool_paths, double offset_distance)
...
+
+  for (std::size_t tool_path_index = 0; tool_path_index < tool_paths.size(); ++tool_path_index)
+  {
+    ToolPath tool_path = tool_paths[tool_path_index];
+    // Sanity check - tool path must have exactly one segment
+    if (tool_path.size() != 1)
+    {
+      ROS_ERROR_STREAM("Tool path " << tool_path_index << " contains " << tool_path.size() <<
+                       " segments. Expected exactly 1.");
+      // TODO: throw an exception
```
Comment (DavidMerzJr) [inline tool_path_planner/src/utilities.cpp]: Again, wondering why we MUST have exactly one segment.

File: noether_msgs/msg/EigenValueEdgeGeneratorConfig.msg [hunk 12]
```diff
@@ -28,5 +28,5 @@ int32 max_intersecting_voxels
 # Any two consecutive points with a shortest distance smaller than this value are merged
 float64 merge_dist
 
-# maximum segment length
-float64 max_segment_length
```
Comment (DavidMerzJr) [inline noether_msgs/msg/EigenValueEdgeGeneratorConfig.msg:32]: Not sure how I feel about removing the previous solution - couldn't they have been left in parallel? Now if we want to go back for any reason (or combine the approaches), we have to edit Noether again.

File: tool_path_planner/CMakeLists.txt [hunk 13]
```diff
@@ -13,6 +13,7 @@ find_package(catkin REQUIRED COMPONENTS
   geometry_msgs
   noether_msgs
   rosconsole
+  roscpp
```
Comment (DavidMerzJr) [inline tool_path_planner/CMakeLists.txt]: We don't need roscpp except for the test_segment_axes.cpp file, which I have thoughts on above.

**Review by DavidMerzJr — CHANGES_REQUESTED** [review]
I found a number of things that had me scratching my head.  First, what looks like a 'development excecutable' has been added to the repository, and would need to be removed before we merge back into main Noether.  Second, a number of situations that will cause segfaults (unguarded 0-lengths) are marked as 'TODO: throw exception'.  Something needs to be done at those locations, and I don't think it should be throwing segfaults.  Finally, I am not sure that this option should have wholesale replaced the previous splitting method - I see no reason why we could not have left both in and let people choose which to use.

File: tool_path_planner/src/test_segment_axes.cpp (hunk 3 above)
Comment (DavidMerzJr) [reply tool_path_planner/src/test_segment_axes.cpp]: Actually, it looks like you have already integrated to utest.cpp, so I am not sure what additional value this file adds.

File: noether_msgs/msg/EigenValueEdgeGeneratorConfig.msg (hunk 12 above)
Comment (colin-lewis-19) [reply noether_msgs/msg/EigenValueEdgeGeneratorConfig.msg:32]: I don't entirely disagree, but the call to remove it came from on high

Comment (DavidMerzJr) [reply noether_msgs/msg/EigenValueEdgeGeneratorConfig.msg:32]: Gotcha.  No shame in acquiescing to on high.

**Comment by DavidMerzJr** [comment]
I resolved the merge conflicts with the online interface, so all you should need to do is `git pull` to pull down the changes, no rebasing required.

**Comment by steviedale** [comment]
@DavidMerzJr Thanks for taking a look at this PR. I agree with you on removing the test node, removing the `using` statements, and handling assumption violations (about the content of the passed in ToolPaths/ToolPathSegment objects) without seg-faulting.

The following commits address these issues:
- d845ba5 --> removed development test node
- e9671a0 --> handle violated assumptions gracefully & remove `using` statements

File: tool_path_planner/src/utilities.cpp (hunk 11 above)
Comment (steviedale) [reply tool_path_planner/src/utilities.cpp]: Discussed offline and addressed in 78b6fc2

File: tool_path_planner/CMakeLists.txt (hunk 13 above)
Comment (steviedale) [reply tool_path_planner/CMakeLists.txt]: Addressed in 3688fa4

File: tool_path_planner/src/utilities.cpp (hunk 4 above)
Comment (colin-lewis-19) [reply tool_path_planner/src/utilities.cpp]: I personally disagree. An empty toolpath should not be provided to an edge path splitter. I think this should throw an exception. Returning an empty toolpath requires the calling code to check for an empty toolpath, which it has already **not** done by passing an empty toolpath into this function.

**Pull request merged**